api/src/main/java/com/lantanagroup/link/api/controller/ReportDataController.java

```
@@ -39,4 +53,64 @@ public void retrieveData(@RequestBody() byte[] content, @PathVariable("type") St
+  public void expungeData() {
+    FhirDataProvider fhirDataProvider = getFhirDataProvider();
+
+    if(dataGovernanceConfig != null) {
+      Bundle censusBundle = fhirDataProvider.getCensusLists();
+      if(censusBundle != null) {
+        for (Bundle.BundleEntryComponent entry : censusBundle.getEntry()) {
+          if(dataGovernanceConfig.getCensusListRetention() != null) {
+            Date lastUpdate = entry.getResource().getMeta().getLastUpdated();
+            expungeData(fhirDataProvider, dataGovernanceConfig.getCensusListRetention(),
+                    lastUpdate, entry.getResource().getId(), entry.getResource().getResourceType().toString());
+          }
+        }
+      }
+
+      Bundle bundle = fhirDataProvider.getAllResources();
+      if(bundle != null) {
+        for(Bundle.BundleEntryComponent entry : bundle.getEntry()) {
+          String tag = entry.getResource().getMeta().getTag().get(0).getCode();
+          Date lastUpdate = entry.getResource().getMeta().getLastUpdated();
+          if((tag.equals(Constants.patientDataTag) && dataGovernanceConfig.getPatientDataRetention() != null) ||
+                  (tag.equals(Constants.reportTag) && dataGovernanceConfig.getReportRetention() != null)) {
+            expungeData(fhirDataProvider, dataGovernanceConfig.getPatientDataRetention(),
+                    lastUpdate, entry.getResource().getId(), entry.getResource().getResourceType().toString());
+          }
+        }
+      }
+    }
+  }
+
+  private void expungeData(FhirDataProvider fhirDataProvider, String retentionPeriod, Date lastDatePosted, String id, String type) {
+    try {
+      Date comp = adjustTime(retentionPeriod, lastDatePosted);
+      Date today = new Date();
+
+      if(today.compareTo(comp) >= 0) {
+        fhirDataProvider.deleteResource(type, id, true);
+        logger.info(type + ": " + id + " has been expunged.");
+      }
+    } catch (DatatypeConfigurationException e) {
+      e.printStackTrace();
```

Maybe better to throw the exception to the caller.